### ISSUE TITLE
Support GCP Service Account Credential Source

### ISSUE THREAD

**Issue by AaronME** [crossplane/crossplane-runtime#281]
### What problem are you facing?
GKE support for Basic Authentication has been deprecated past Kubernetes 1.18. Moving forward, crossplane will need to support authentication based on GCP Service Accounts for GKE clusters.

This configuration will require passing both a properly constructed Kubeconfig and a Google Service Account Credentials Json to the provider.

### How could Crossplane help solve your problem?
Create a GoogleServiceAccount CredentialSource which can source both the kubeconfig and the credentials.json from a secret. This could happen in one of two ways.

Proposal 1 - Allow sourcing credentials from distinct secrets. 
```yaml
---
apiVersion: helm.crossplane.io/v1alpha1
kind: ProviderConfig
metadata:
  name: gke
spec:
  credentials:
    source: GKEServiceAccount
    kubeconfig:
      secretRef:
        name: secret-name-one
        key: kubeconfig
    credentials.json:
      secretRef:
        name: secret-name-two
        key: credentials.json
```

Proposal 2 - Create a single secret with multiple fields:
```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: gke-creds
data:
  kubeconfig: ...
  credentials.json: ...
type: Opaque
---
apiVersion: helm.crossplane.io/v1alpha1
kind: ProviderConfig
metadata:
  name: gke
spec:
  credentials:
    source: GKEServiceAccount
    secretRef:
      name: gke-creds
```

**Comment by AaronME** [comment]
Possible duplicate of https://github.com/crossplane/crossplane-runtime/issues/215

Since the "GKEServiceAccount" is actually just a reference to two different secret keys (in one or more secrets) we might be better off supporting a multi-secret CredentialSource which could be leveraged by more implementations. Immediate needs include:

1. Alibaba
2. GKE Cluster Authentication
3. EKS IRSA Support

**Comment by negz** [comment]
One secret feels like a nicer UX on the surface, but I suspect two secrets would make mixing and matching stuff via a `Composition` easier. I could imagine for example creating a `GKECluster` and a `ServiceAccount` that wrote a `kubeconfig` and a `credentials.json` to their respective secrets, then using those two to go deploy a Helm `Release` to the cluster using the service accounts credentials all in one XR.

**Comment by negz** [comment]
> Since the "GKEServiceAccount" is actually just a reference to two different secret keys (in one or more secrets) we might be better off supporting a multi-secret CredentialSource which could be leveraged by more implementations.

Hrm, the code that consumes this `ProviderConfig` would need to know which secret key was the `kubeconfig` and which was the `credentials.json` right? I kind of lean toward a special purpose credentials source, though I can see that would probably blow out into lots of special purpose sources.

**Comment by turkenh** [comment]
I believe this issue is gaining more importance as Kubernetes versions in GKE moves up and `v1.18` is becoming an old version.

**Comment by negz** [comment]
https://github.com/kubernetes/client-go/blob/v0.22.1/plugin/pkg/client/auth/gcp/gcp.go#L122
https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins

I just took a quick look into cleaning up https://github.com/crossplane-contrib/provider-helm/pull/98 and getting it added to crossplane-runtime, because I'm potentially blocked on this now too with my scale testing. While doing that I noticed the line above.

It seems like at least the GCP and Azure auth providers are now deprecated in favour of shelling out to a command (e.g. `gcloud`) to grab credentials. That said, it doesn't look like the actual `rest.AuthProvider` pattern is deprecated. 🤔

**Comment by negz** [comment]
So I found myself blocked on this today and thought I'd "quickly" update @AaronME's https://github.com/crossplane-contrib/provider-helm/pull/98 PR with some of the ideas we've discussed in this issue. Four PhDs in GKE authentication later, I ended up just opening https://github.com/crossplane-contrib/provider-helm/pull/109 which is a slightly polished version of @AaronME's PR. 🙃 

My rationale for going back to the "do it in each provider" approach is:

* It's tricky to land on an API that feels sufficiently generic to warrant inclusion in crossplane-runtime.
* The actual code that turns Google Application Credentials JSON bytes into a `*rest.Config` is pretty tiny.
* Maybe if we build this a few different times we'll discover a way to move it into crossplane-runtime later?

When I tried to to model this API in crossplane-runtime but I found it felt overly specific and out-of-place if I went with the simple `source: GKEServiceAccount` route (that presumably would actually take two credentials - a `kubeconfig` and a `credentials.json`.  On the other hand, it felt overly generic to try to model it as an array of credentials. I think in order to avoid a breaking API change we'd end up with something like:

```yaml
apiVersion: helm.crossplane.io/v1beta1
kind: ProviderConfig
metadata:
  name: default
spec:
  credentials:
    source: Multiple
    multipleSources:
    - purpose: Kubeconfig
       source: Secret
       secretRef:
         namespace: crossplane-system
         name: cluster-creds
         key: kubeconfig
    - purpose: GoogleApplicationCredentials
       source: Secret
       secretRef:
         namespace: crossplane-system
         name: gcp-creds
         key: credentials.json
```

The purposes of each secret would presumably be meaningful to each different provider and hard to generalise.

**Comment by negz** [comment]
We ended up merging https://github.com/crossplane-contrib/provider-helm/pull/109, and I've replicated that pattern in https://github.com/crossplane-contrib/provider-kubernetes/pull/10. Once the latter PR is merged I'm going to consider this closed for the time being - we can reassess whether this code (at least `gke.go`?) should be added to controller-runtime if it we see it propagating to more providers.
